src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,41 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private Node<T> top;
+    private int size;
 
+    public void push(T value) {
+        top = new Node<>(value, top);
+        size++;
     }
 
     public T peek() {
-        return null;
+        if (size == 0) {
+            throw new EmptyStackException();
+        }
+        return top.value;
     }
 
     public T pop() {
-        return null;
+        T poppedValue = peek();
+        top = top.previous;
+        size--;
+        return poppedValue;
     }
 
     public int size() {
-        return 0;
+        return size;
+    }
+
+    private static class Node<T> {
+        private final T value;
+        private final Node<T> previous;
```

Comment:
```suggestion
        private final Node<T> next;
```

Reply:
For me more logical is previous. I think it is not matter much